vendor/github.com/openshift/api/apiserver/v1/types.go

```diff
@@ -0,0 +1,83 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CriticalResource prevents the .spec.provider from being deleted until the .spec.criteria are satisfied.
+// Because the ability to block deletion of a namespace scoped resource implies the ability to prevent namespaces
+// from being cleaned up, these are namespace scoped, but they have a higher degree of privilege required for mutation
+// than namespace admin.
+// TODO in a future release we can consider making the permissions broader, but because criteria span namespaces,
+//  it seems unlikely that we can broaden the permissions.
+type CriticalResource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   CriticalResourceSpec   `json:"spec"`
+	Status CriticalResourceStatus `json:"status"`
+}
+
+type CriticalResourceSpec struct {
+	// provider indicates the resource which provides critical functionality.  It must be in the same namespace as the
```

Comment:
Just a little bit of bikeshedding, I think `resource` might be an easier name for folks to understand

Reply:
+1

Provider is not intuitive. Without reading here you won't understand what a provider is for a critical resource. Just call it "resource", or even inline.